cpp/src/arrow/chunked_array.cc

```diff
@@ -147,13 +148,15 @@ bool ChunkedArray::ApproxEquals(const ChunkedArray& other,
 }
 
 Result<std::shared_ptr<Scalar>> ChunkedArray::GetScalar(int64_t index) const {
-  for (const auto& chunk : chunks_) {
-    if (index < chunk->length()) {
-      return chunk->GetScalar(index);
-    }
-    index -= chunk->length();
+  if (!chunk_resolver_) {
+    chunk_resolver_ = internal::make_unique<internal::ChunkResolver>(chunks_);
+  }
+  const auto loc = chunk_resolver_->Resolve(index);
+  if (loc.chunk_index >= static_cast<int64_t>(chunks_.size())) {
+    return Status::IndexError("tried to refer to chunk ", loc.chunk_index,
+                              " but ChunkedArray is only ", chunks_.size(), " long");
```

Comment:
That's a weird error message. The cause is probably that the index passed to `GetScalar` is out of bounds, why not simply say that?

Reply:
I agree, but I simply used an [existing `Status::IndexError` comment found in `array_base.cc`](https://github.com/apache/arrow/blob/master/cpp/src/arrow/array/array_base.cc#L168-L169).